File: pompo-backend/app/schemas/webhook.py
"""Public webhook API contracts."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class WebhookIngestResponse(BaseModel):
    status: str = Field(description="Acknowledgement status for the provider callback.")
    event_id: str = Field(description="POMPO public webhook event identifier.")
    duplicate: bool = Field(default=False, description="True when the provider event was already received.")


class WebhookEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    public_identifier: str
    provider_id: uuid.UUID
    provider_event_id: str
    transaction_id: uuid.UUID | None
    payment_attempt_id: uuid.UUID | None
    event_type: str
    event_version: str | None
    payment_reference: str | None
    provider_transaction_reference: str | None
    received_at: datetime
    processed_at: datetime | None
    processing_status: str
    processing_attempts: int
    signature_verified: bool
    timestamp_validated: bool
    failure_category: str | None
    failure_code: str | None
    payload: dict

File: pompo-frontend/src/lib/types/webhook.ts
export interface WebhookEventRecord {
  id: string;
  public_identifier: string;
  provider_id: string;
  provider_event_id: string;
  transaction_id: string | null;
  payment_attempt_id: string | null;
  event_type: string;
  event_version: string | null;
  payment_reference: string | null;
  provider_transaction_reference: string | null;
  received_at: string;
  processed_at: string | null;
  processing_status: string;
  processing_attempts: number;
  signature_verified: boolean;
  timestamp_validated: boolean;
  failure_category: string | null;
  failure_code: string | null;
  payload: Record<string, unknown>;
}
